File: mergeSort.js
let lista = [1,2,5,45,36,4,3653,4,23,6];

// push() : Adiciona um novo elemento ao final do array
// Operador condicional terciário : condição ? expressão1 : expressão2
// shift() : Remove o primeiro elemento do array e retorna ele.
//Math.floor() : Arredonda o float para baixo e o transforma em inteiro.
//slice() : Retorna um subarray a partir do array informado.

function merge (a,b) {
    var result = [];
    while (a.length >0 && b.length >0)
        result.push(a[0] < b[0]? a.shift() : b.shift());
    return result.concat(a.length? a : b);
}

function mergeSort (arr) {
    if (arr.length < 2) return arr;
    var mid = Math.floor(arr.length /2);
    var subLeft = mergeSort(arr.slice(0,mid));
    var subRight = mergeSort(arr.slice(mid));
    return merge(subLeft, subRight);
}

console.log(mergeSort(lista));

File: mergeSort.py
'''

len() : Tamanho do array

Função lambda : 

Ex.: soma = lambda x, y: x + y
print(soma(2, 3))

'''

def merge(a,b,compare):
	result = [] 
	i,j = 0,0
	while (i < len(a) and j < len(b)):
		if compare(a[i],b[j]):
			result.append(a[i])
			i += 1
		else:
			result.append(b[j])
			j += 1
	while (i < len(a)):
		result.append(a[i])
		i += 1
	while (j < len(b)):
		result.append(b[j])
		j += 1
	return result

def merge_sort(arr, compare = lambda x, y: x < y):
	if len(arr) < 2:
		return arr[:] # Shallow Copy
	else:
		middle = len(arr) // 2 
		a = merge_sort(arr[:middle], compare)
		b = merge_sort(arr[middle:], compare)
		return merge(a, b, compare) 

arr = [1,13,2,52,342,4,13,1,31,2]
print(merge_sort(arr))
